scrapy/utils/item.py

```diff
@@ -0,0 +1,104 @@
+from collections.abc import MutableMapping
+
+from scrapy.item import BaseItem
+
+
+def _is_dataclass_instance(obj):
+    """
+    Return True if *obj* is a dataclass object, False otherwise.
+    Taken from https://docs.python.org/3/library/dataclasses.html#dataclasses.is_dataclass.
+    """
+    try:
+        from dataclasses import is_dataclass
+    except ImportError:
+        return False
+    else:
+        return is_dataclass(obj) and not isinstance(obj, type)
+
+
+def is_item_like(obj):
+    """
+    Return True if *obj* is considered a Scrapy *item*, False otherwise.
+
+    An object is considered an *item* if it is:
+    - a scrapy.item.BaseItem or dict instance (or any subclass)
+    - a dataclass object
+    """
+    return isinstance(obj, (BaseItem, dict)) or _is_dataclass_instance(obj)
+
+
+class ItemAdapter(MutableMapping):
```

Comment:
I don't want to derail the PR, but I'm wondering now if we should create a separate package with ItemAdapter. A package which provides  an uniform dict-like access to various sources of data: scrapy Items, dicts, dataclasses, probably attr.s.

Where this idea comes from: scrapy.loader is being separated to a different package, and after this PR scrapy.loader depends on ItemAdapter.

Reply:
/cc @ejulio

Reply:
@kmike That makes sense. I can create a repo under my account and start working on it, we can transfer it later. Does that sound good?

Reply:
@elacuesta sure, that'd be perfect!

Reply:
Working on this at https://github.com/elacuesta/scrapy-itemadapter

Reply:
@Gallaecio @kmike `itemadapter 0.0.1` is on PyPI, I can start updating this PR. One thing I think we might want to discuss is whether or not we want to import and use directly from the package, or maybe do something similar to what we do with Parsel. An argument in favour of this would be restricting the usage to `Item` objects in Scrapy, since I don't think we should do that in `itemadapter` (because, for the purposes of the library, any `MutableMapping` does work). What do you think?

Reply:
My main concern is the handling of `{'fields': '…'}` in non-`Item` objects. I’m not sure it is safe to assume that if the attribute exists we can treat the object as a Scrapy `Item`. Best case scenario: a puzzling `TypeError` exception is raised. Worst case scenario: thinks work in unintended ways.

Reply:
@elacuesta I agree with @Gallaecio; I think scrapy.Item should be handled not in the same way as MutableMapping in itemadapter package, and it is better to do a conditional import of scrapy.Item in itemadapter.

So in a way, it is a circular dependency (not necessarily circular import). itemadapter can work without Scrapy, but supports Scrapy Items, just like it supports dataclasses or attr.s objects.

For me it looks like a separate wrapper is not required here; we can use it directly.

Reply:
I think we would lose generality in `itemadapter` if we do an explicit check for Scrapy items. The way I see it is:
* We check for Scrapy items, and decide to remove support for meta in general for mutable mappings. We're disallowing something that we can support and could be useful for users
* We check for Scrapy items, and decide to keep support for meta in general for mutable mappings. The code for getting the fields is the same for items as it is for mutable mappings, it ends up being the same as not checking in the first place

I think one alternative is to restrict things in the Scrapy side, i.e.:

```python
# scrapy/utils/item.py

from scrapy.item import Item

from itemadapter.adapter import ItemAdapter as BaseItemAdapter
from itemadapter.utils import is_attrs_instance, is_dataclass_instance


def is_item(obj):
    return isinstance(obj, Item) or is_dataclass_instance(obj) or is_attrs_instance(obj)


class ItemAdapter(BaseItemAdapter):
    def __init__(self, item):
        if not is_item(item):
            raise TypeError("Expected a valid item, got %r instead: %s" % (type(item), item))
        super().__init__(item)
```

Reply:
By “support for meta” do you mean the 2 methods to get field metadata? Wouldn’t it make sense _not_ to support that in `MutableMapping` given `MutableMapping` does not support defining metadata?

Otherwise we are saying that, in mappings used with this library, the `fields` key is reserved for field metadata. If field metadata support is desired, I think something other than a simple `dict`, `MutableMapping` or `Mapping` should be used.

Reply:
Not the key but the _attribute_, i.e. `obj.fields` not `obj["fields"]`. My reasoning was "if we support arbitrary dicts, why not support arbitrary dicts which also happen to define a `fields` attribute?"
Anyway, this probably will not be an issue in practice, as you say, I imagine users using other types if they need metadata, I just thought it would be nice to allow and not difficult to maintain. I can remove that in itemadapter, it's already released but it's not like many people are using it.

Reply:
> Not the key but the attribute, i.e. obj.fields not obj["fields"]

I did not realize this :facepalm: . In that case I think it is OK to support it. If we do, though, the documentation should definitely cover this, and explain how to implement such an attribute.

Reply:
Hey! Currently in Scrapy it works like that:

* dict is an item, meta is not supported
* Item is an item, meta is supported

After this PR, before itemadapter package:
* dataclass instance is an item, meta is not supported (for now; we could support https://docs.python.org/3.9/library/typing.html#typing.Annotated in future)

In the itemadapter, non-controversial:
* attr.s instance is an item, meta is supported (though is the support implemented?)

The thing we're discussing, a controversial change in itemadapter:
* arbitrary MutableMapping is an item
* for arbitrary MutableMapping meta is supported if it has "fields" attribute.

I think the last part may require a separate discussion - do we allow any arbitrary MutableMapping as an item, or not. If we allow arbitrary MutableMapping, how to annotate fields. This change is much larger than adding dataclass support. I recall initially we decided to be very explicit, and so Scrapy is checking just for `dict` now, not for any MutableMapping - it'd be easy to add support for arbitrary MutableMapping in future, but it'd be hard to remove it, if we find some issue with that. It may have a merit, and use cases, and could be non-problematic, but I think it deserves a separate discussion.

What do you think about focusing on non-controversial, low risk changes first, to get everything merged? I.e. rollback / move to PR controversial stiff. And then we can discuss MutableMapping thing separately.

Reply:
The difference is that for dicts, scrapy.Item, dataclasses and attr.s instances we know what we're working with; for MutableMapping we don't, as all kinds af object in Python ecosystem may be implementing it. That's why I see it as a higher risk change.

Reply:
Sounds fine to me, I removed meta for MutableMappings already, I'll remove MutableMappings altogether now, it'd be easy to add it back if we decide to.

Reply:
Done in [v0.0.5](https://github.com/elacuesta/itemadapter/releases/tag/v0.0.5).
One thing I wonder is: will there be any issues when running `itemadapter`'s tests (which install Scrapy to check for `scrapy.item.Item`) if Scrapy setup requires `ìtemadapter`?